src/main/webapp/app/course/tutorial-groups/tutorial-groups-management/tutorial-free-periods/crud/tutorial-free-period-form/tutorial-group-free-period-form.component.ts

```diff
@@ -1,44 +1,150 @@
     @Input()
     formData: TutorialGroupFreePeriodFormData = {
-        date: undefined,
+        startDate: undefined,
+        endDate: undefined,
+        startTime: undefined,
+        endTime: undefined,
         reason: undefined,
     };
 
     @Input() isEditMode = false;
 
     @Input() timeZone: string;
 
-    @Output() formSubmitted: EventEmitter<TutorialGroupSessionFormData> = new EventEmitter<TutorialGroupSessionFormData>();
+    @Output() formSubmitted: EventEmitter<TutorialGroupFreePeriodFormData> = new EventEmitter<TutorialGroupFreePeriodFormData>();
 
     faCalendarAlt = faCalendarAlt;
 
     form: FormGroup;
 
-    get dateControl() {
-        return this.form.get('date');
+    protected timeFrame = TimeFrame.Day;
+
+    protected readonly TimeFrame = TimeFrame;
+
+    // Todo: TimeFrame getter/setter. Reset endDate when switching back to single Day
+    setTimeFrame(timeFrame: TimeFrame) {
+        if (timeFrame === TimeFrame.Day) {
+            this.resetDateControl('endDate');
+            this.resetDateControl('endTime');
+            this.resetDateControl('startTime');
+        } else if (timeFrame == TimeFrame.Period) {
+            this.resetDateControl('startTime');
+            this.resetDateControl('endTime');
+        } else if (timeFrame == TimeFrame.PeriodWithinDay) {
+            this.resetDateControl('endDate');
+        }
+        this.timeFrame = timeFrame;
+    }
+
+    private resetDateControl(controlName: string) {
+        if (this.form.get(controlName)) {
+            this.form.get(controlName)?.reset();
+            this.form.get(controlName)?.markAsUntouched();
+        }
+    }
+
+    get isStartBeforeEnd() {
+        if (this.timeFrame == TimeFrame.PeriodWithinDay && this.endTimeControl && this.startTimeControl) {
+            return this.endTimeControl.value > this.startTimeControl.value;
+        } else if (this.timeFrame == TimeFrame.Period && this.endDateControl && this.startDateControl) {
+            return this.endDateControl.value > this.startDateControl.value;
+        } else {
+            return true;
+        }
+    }
+
+    get isFreeDay(): boolean {
+        return this.timeFrame == TimeFrame.Day;
+    }
+
+    get isFreePeriod(): boolean {
+        return this.timeFrame == TimeFrame.Period;
+    }
+
+    get isFreePeriodWithinDay(): boolean {
+        return this.timeFrame == TimeFrame.PeriodWithinDay;
+    }
+
+    get timeFrameControl(): TimeFrame {
+        return this.timeFrame;
+    }
+
+    get startDateControl() {
+        return this.form.get('startDate');
+    }
+
+    get endDateControl() {
+        return this.form.get('endDate');
+    }
+
+    get startTimeControl() {
+        return this.form.get('startTime');
+    }
+
+    get endTimeControl() {
+        return this.form.get('endTime');
     }
 
     get reasonControl() {
         return this.form.get('reason');
     }
 
     get isSubmitPossible() {
```

Comment:
Add comments which explain the logic of this functionality.

Reply:
I greatly simplified this function.